suite/tests/linux/mangle_pauth.c

```diff
@@ -0,0 +1,384 @@
+            M(blraa);        \
+            M(blrab);        \
+        } while (0)
+
+#    define NON_FAULT_TEST(instr_name)            \
+        LOG("Non-fault test: " #instr_name "\n"); \
+        test_##instr_name(false)
+
+    FOR_EACH_TEST(NON_FAULT_TEST);
+
+    intercept_signal(SIGSEGV, (handler_3_t)&handle_signal, false);
+    intercept_signal(SIGILL, (handler_3_t)&handle_signal, false);
+
+#    define FAULT_TEST(instr_name)                                                     \
+        LOG("Fault test: " #instr_name "\n");                                          \
+        switch (SIGSETJMP(mark)) {                                                     \
+        default:                                                                       \
+            test_##instr_name(true);                                                   \
+            print(#instr_name " fault test failed: No fault\n");                       \
+            break;                                                                     \
+        case TEST_PC_MISMATCH: print(#instr_name " fault test failed: PC mismatch\n"); \
+        case TEST_PASS: break;                                                         \
+        }
+    FOR_EACH_TEST(FAULT_TEST);
+
+    print("Test complete\n");
+
+    return 0;
+}
+
+#else /* asm code *************************************************************/
+/* clang-format off */
+#    include "asm_defines.asm"
+START_FILE
+
+.arch armv8.3-a
+
+DECL_EXTERN(branch_instr_addr)
+DECL_EXTERN(branch_target_addr)
+DECL_EXTERN(dummy_func)
+
+#define PASTE(a, b) a##b
+
+/* Use inside a function to declare or reference a function-local label.
+ * Expects FUNCNAME to be defined.
+ */
+#define _LOCAL_LABEL(label, unique_id) PASTE(unique_id, label)
+#define LOCAL_LABEL(label) _LOCAL_LABEL(label, FUNCNAME)
+
+/* Write a value to the global branch_target_addr variable. */
+#define SAVE_BRANCH_TARGET_ADDR(addr_reg, tmp_reg)                    \
+        AARCH64_ADRP_GOT(GLOBAL_REF(branch_target_addr), tmp_reg) @N@ \
+        str     addr_reg, [tmp_reg]
+
+/* Write the address of a label to the global branch_instr_addr variable. */
+#define SAVE_BRANCH_INSTR_ADDR(label, tmp_reg1, tmp_reg2)             \
+        AARCH64_ADRP_GOT(GLOBAL_REF(branch_instr_addr), tmp_reg1) @N@ \
+        adr     tmp_reg2, label                                   @N@ \
+        str     tmp_reg2, [tmp_reg1]
+
+/* Give some names to some of the registers we need in the tests. */
+#define TRIGGER_FAULT_ARG_REG x0
+#define BRANCH_TARGET_REG x1
+#define TMP1_REG x2
+#define TMP2_REG x3
+#define MODIFIER_REG x4
+
+#define FUNCNAME test_retaa
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        SAVE_BRANCH_TARGET_ADDR(x30, TMP1_REG)
+        SAVE_BRANCH_INSTR_ADDR(LOCAL_LABEL(branch_instr), TMP1_REG, TMP2_REG)
+        paciasp
+        /* target_addr = signed_addr ^ (trigger_fault << 53); */
+        eor     x30, x30, TRIGGER_FAULT_ARG_REG, lsl 53
```

Comment:
(To make sure I understand: we enter this function with a regular branch-and-link.  We save the (plain) return address and the PC of our RETAA into global vars for the signal handler to access.  We then encode the return address, flip bit 53 if we're asked to generate a fault, and ask to decode and return.)

Reply:
Yes that's right.